pymc3/step_methods/hmc/nuts.py

```diff
@@ -352,24 +308,44 @@ def extend(self, direction):
 
         return diverging, turning
 
-    def _build_subtree(self, left, depth, epsilon):
-        if depth == 0:
+    def _single_step(self, left, epsilon):
+        """Perform a leapfrog step and handle error cases."""
+        try:
             right = self.leapfrog(left.q, left.p, left.q_grad, epsilon)
+        except linalg.LinAlgError as err:
+            error_msg = "LinAlgError during leapfrog step."
+            error = err
+        except ValueError as err:
+            # Raised by many scipy.linalg functions
+            scipy_msg = "array must not contain infs or nans"
+            if len(err.args) > 0 and scipy_msg in err.args[0].lower():
+                error_msg = "Infs or nans in scipy.linalg during leapfrog step."
+                error = err
+            else:
+                raise
+        else:
             right = Edge(*right)
             energy_change = right.energy - self.start_energy
             if np.isnan(energy_change):
                 energy_change = np.inf
 
             if np.abs(energy_change) > np.abs(self.max_energy_change):
                 self.max_energy_change = energy_change
-            p_accept = min(1, np.exp(-energy_change))
-
-            log_size = -energy_change
-            diverging = energy_change > self.Emax
+            if np.abs(energy_change) < self.Emax:
+                p_accept = min(1, np.exp(-energy_change))
+                log_size = -energy_change
+                proposal = Proposal(right.q, right.energy, p_accept)
+                tree = Subtree(right, right, right.p, proposal, log_size, p_accept, 1)
+                return tree, False, False
+            else:
+                error_msg = "Bad energy after leapfrog step."
```

Comment:
The step size should be adapted already when you see this message. Increasing `target_accept` is I think the right thing (this will change the step size adaptation to use smaller step sizes). But otherwise I agree.